### PR TITLE
refactor: fix clang-tidy reports

### PR DESCRIPTION
**Pull request by JakubFornadel**
## Purpose

This basically code clean-up with:

- performance fixes
- move semantics fixes
- etc...

### CODE REVIEW

File: libraries/common/include/common/lazy.hpp [hunk 1]
```diff
@@ -14,6 +14,7 @@ using std::remove_reference;
 using std::size_t;
 using std::unique_lock;
 
+// TODO: Oleh's legacy of bad and ridiculously complicated code - clean up this class !!!
```
Comment (gaoqi8649) [inline libraries/common/include/common/lazy.hpp:17]: Need create issue for this and put number after TODO

File: libraries/config/src/config.cpp [hunk 2]
```diff
@@ -270,6 +270,10 @@ void NetworkConfig::validate() const {
                           std::to_string(MAX_PACKETS_PROCESSING_THREADS_NUM) + "]");
   }
 
+  if (network_transaction_interval < 0) {
```
Comment (gaoqi8649) [inline libraries/config/src/config.cpp:273]: The type of network_transaction_interval is uint16_t, that will not < 0

File: libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp [hunk 3]
```diff
@@ -7,7 +7,7 @@
 namespace taraxa {
 
 class FullNode;
-class PbftManager;
+class VoteManager;
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp:10]: Is this a typo?

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 4]
```diff
@@ -288,8 +288,8 @@ DagManager::DagManager(blk_hash_t const &genesis, addr_t node_addr, std::shared_
       log_time_(log_time) {
   LOG_OBJECTS_CREATE("DAGMGR");
   if (auto ret = getLatestPivotAndTips(); ret) {
-    frontier_.pivot = std::move(ret->first);
-    for (auto const &t : ret->second) {
+    frontier_.pivot = ret->first;
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/src/dag/dag.cpp:291]: Why not std::move(ret->first) here?

File: libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp [hunk 5]
```diff
@@ -115,8 +115,8 @@ void SortitionParamsManager::pbftBlockPushed(const SyncBlock& block, DbStorage::
 
     if (non_empty_pbft_chain_size % config_.changing_interval == 0) {
       const auto params_change = calculateChange(period);
-      params_changes_.push_back(params_change);
-      db_->saveSortitionParamsChange(period, std::move(params_change), batch);
+      params_changes_.push_back(std::move(params_change));
+      db_->saveSortitionParamsChange(period, params_change, batch);
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp:119]: After std::move(params_change), cannot use params_change again

File: libraries/core_libs/consensus/src/final_chain/final_chain.cpp [hunk 6]
```diff
@@ -105,7 +105,7 @@ class FinalChainImpl final : public FinalChain {
           continue;
         }
         // Non-executed trxs
-        auto const& trx = new_blk.transactions[i];
+        auto& trx = new_blk.transactions[i];
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/src/final_chain/final_chain.cpp:108]: const auto trx? Since std::move(trx) later

**Review by gaoqi8649 — APPROVED** [review]

File: libraries/config/src/config.cpp (hunk 2 above)
Comment (JakubFornadel) [reply libraries/config/src/config.cpp:273]: true, the condition is supposed to be <= 0, thanks

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 4 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/dag/dag.cpp:291]: because some types dont use internally pointers, like form example blkh_hash_t so it does not make sense to move as nothing is gained

File: libraries/core_libs/consensus/src/final_chain/final_chain.cpp (hunk 6 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/final_chain/final_chain.cpp:108]: your suggestion is exactly the problem here - you cant move anything that is const, copy is used in such case...

File: libraries/cli/include/cli/default_config.hpp [hunk 7]
```diff
@@ -1,6 +1,10 @@
+#pragma once
+
+#include "string_view"
```
Comment (MatusKysel) [inline libraries/cli/include/cli/default_config.hpp:3]: <string_view>

File: libraries/cli/include/cli/devnet_config.hpp [hunk 8]
```diff
@@ -1,6 +1,10 @@
+#pragma once
+
+#include "string_view"
```
Comment (MatusKysel) [inline libraries/cli/include/cli/devnet_config.hpp:3]: <string_view>

File: libraries/cli/include/cli/mainnet_config.hpp [hunk 9]
```diff
@@ -1,6 +1,10 @@
+#pragma once
+
+#include "string_view"
```
Comment (MatusKysel) [inline libraries/cli/include/cli/mainnet_config.hpp:3]: <string_view>

File: libraries/cli/include/cli/testnet_config.hpp [hunk 10]
```diff
@@ -1,6 +1,10 @@
+#pragma once
+
+#include "string_view"
```
Comment (MatusKysel) [inline libraries/cli/include/cli/testnet_config.hpp:3]: same here

File: libraries/common/include/common/default_construct_copyable_movable.hpp [hunk 11]
```diff
@@ -14,13 +14,14 @@ struct DefaultConstructCopyableMovable {
   T val;
 
   DefaultConstructCopyableMovable() = default;
-  DefaultConstructCopyableMovable(T val) : val(std::move(val)) {}
+  ~DefaultConstructCopyableMovable() = default;
```
Comment (MatusKysel) [inline libraries/common/include/common/default_construct_copyable_movable.hpp:17]: why does it need destructor?

File: libraries/core_libs/consensus/include/dag/dag.hpp [hunk 12]
```diff
@@ -63,8 +63,15 @@ class Dag {
...
 
   friend DagManager;
+
   explicit Dag(blk_hash_t const &genesis, addr_t node_addr);
   virtual ~Dag() = default;
+
+  Dag(const Dag &) = default;
+  Dag(Dag &&) = default;
+  Dag &operator=(const Dag &) = default;
+  Dag &operator=(Dag &&) = default;
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/include/dag/dag.hpp:73]: why is this needed? shouldn't be constructed by default, without this definitions?

File: libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp [hunk 13]
```diff
@@ -7,26 +7,32 @@
...
 
-  virtual bool is_nonce_stale(const addr_t& addr, const trx_nonce_t& nonce) const = 0;
+  virtual bool is_nonce_stale(const addr_t &addr, const trx_nonce_t &nonce) const = 0;
 
   struct TransactionInfo {
     addr_t sender;
     trx_nonce_t nonce = 0;
   };
-  virtual void update(DB::Batch&, uint64_t, util::RangeView<TransactionInfo> const&) = 0;
+  virtual void update(DB::Batch &, uint64_t, util::RangeView<TransactionInfo> const &) = 0;
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:29]: if you already touching this you should change the order of const

File: libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp [hunk 14]
```diff
@@ -7,26 +7,32 @@
...
 };
 
 std::unique_ptr<ReplayProtectionService> NewReplayProtectionService(ReplayProtectionService::Config,
                                                                     std::shared_ptr<DB>);
 
-Json::Value enc_json(ReplayProtectionService::Config const&);
-void dec_json(Json::Value const&, ReplayProtectionService::Config&);
+Json::Value enc_json(ReplayProtectionService::Config const &);
+void dec_json(Json::Value const &, ReplayProtectionService::Config &);
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:36]: same here

File: libraries/core_libs/consensus/include/final_chain/state_api_data.hpp [hunk 15]
```diff
@@ -14,13 +14,11 @@ namespace taraxa::state_api {
 
 struct TaraxaEVMError : std::runtime_error {
   std::string const type;
-  TaraxaEVMError(std::string type, std::string msg);
-  ~TaraxaEVMError() throw();
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/include/final_chain/state_api_data.hpp:18]: are you sure about this? Did you try to cause this exception?

File: libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp [hunk 16]
```diff
@@ -116,7 +116,7 @@ void SortitionParamsManager::pbftBlockPushed(const SyncBlock& block, DbStorage::
     if (non_empty_pbft_chain_size % config_.changing_interval == 0) {
       const auto params_change = calculateChange(period);
       params_changes_.push_back(params_change);
-      db_->saveSortitionParamsChange(period, std::move(params_change), batch);
+      db_->saveSortitionParamsChange(period, params_change, batch);
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp:119]: I think it would be better to change it so we can use move here. If I have check where it is used, it should not be a problem

File: libraries/core_libs/consensus/src/pbft/block_proposer.cpp [hunk 17]
```diff
@@ -206,16 +206,13 @@ void BlockProposer::proposeBlock(DagFrontier&& frontier, level_t level, SharedTr
...
-
-  auto now = getCurrentTimeMilliSeconds();
-  LOG(log_time_) << "Propose block " << blk.getHash() << " at: " << now << " ,trxs: " << blk.getTrxs()
-                 << " , tips: " << blk.getTips().size();
+  DagBlock blk(frontier.pivot, level, std::move(frontier.tips), std::move(trx_hashes), std::move(vdf), node_sk_);
   LOG(log_nf_) << "Add proposed DAG block " << blk.getHash() << ", pivot " << blk.getPivot() << " , number of trx ("
                << blk.getTrxs().size() << ")";
+
+  dag_mgr_->addDagBlock(std::move(blk), std::move(trxs), true);
+  dag_blk_mgr_->markDagBlockAsSeen(blk);
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/block_proposer.cpp:214]: ? first move then mark same variable?

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 18]
```diff
@@ -449,7 +448,7 @@ void DagManager::addDagBlock(DagBlock const &blk, SharedTransactions &&trxs, boo
     if (save) {
       block_verified_.emit(blk);
       if (auto net = network_.lock()) {
-        net->onNewBlockVerified(blk, proposed, std::move(trxs));
+        net->onNewBlockVerified(std::move(blk), proposed, std::move(trxs));
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:451]: `blk` is used later in log

File: libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp (hunk 13 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:29]: I am not, clang-format did format it automatically, which is weird that it did not format it before

File: libraries/core_libs/consensus/include/final_chain/state_api_data.hpp (hunk 15 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/final_chain/state_api_data.hpp:18]: this throw() is not actual exception throwing, it tells the compiler that the destructor will not throw... Automatically generated one does the same thing

File: libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp (hunk 16 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp:119]: - in this concrete line we would not benefit from move as saveSortitionParamsChange just takes the rlp so if you take it from rvale&& or const lvalue& does not make much difference
- + SortitionParamsChange class is primitive type, which means it does not contain any data that would use pointer so moing this object always does a copy anyway

**Review by MatusKysel — COMMENTED** [review]
Did you run analysis after your changes? Because I see some use after move and that's really dangerous

File: libraries/common/include/common/default_construct_copyable_movable.hpp (hunk 11 above)
Comment (JakubFornadel) [reply libraries/common/include/common/default_construct_copyable_movable.hpp:17]: the rule is that as soon as you declare any of the 5 ctors/operators, you should declare all of them

File: libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp (hunk 13 above)
Comment (MatusKysel) [reply libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:29]: maybe you did update your clang suite?

Comment (MatusKysel) [reply libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:29]: new clang-format new rules?

File: libraries/config/src/config.cpp [hunk 19]
```diff
@@ -270,6 +270,10 @@ void NetworkConfig::validate() const {
                           std::to_string(MAX_PACKETS_PROCESSING_THREADS_NUM) + "]");
   }
 
+  if (network_transaction_interval == 0) {
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply libraries/config/src/config.cpp:273]: yes, and we did not use it in any test, nor any network

**Comment by JakubFornadel** [comment]
> Did you run analysis after your changes? Because I see some use after move and that's really dangerous

I am running it now but it takes ages so it is not finished yet

File: libraries/core_libs/network/src/tarcap/packets_handlers/transaction_packet_handler.cpp [hunk 20]
```diff
@@ -61,68 +59,68 @@ inline void TransactionPacketHandler::process(const PacketData &packet_data, con
...
+    // incorrectly markTransactionAsKnown
+    if (!peer.second->syncing_) {
+      peers_to_log += peer.first.abridged();
+      for (auto const &trx : transactions) {
+        auto trx_hash = trx->getHash();
+        if (peer.second->isTransactionKnown(trx_hash)) {
+          continue;
         }
+
+        if (transactions_to_send[peer.first].size() > MAX_TRANSACTIONS_IN_PACKET) {
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/packets_handlers/transaction_packet_handler.cpp:106]: yes we still need this - we want to limit all packets sizes that could be too big due to network delays, long processing time etc...

File: tests/full_node_test.cpp [hunk 21]
```diff
@@ -757,7 +757,7 @@ TEST_F(FullNodeTest, insert_anchor_and_compute_order) {
   g_mock_dag0 = samples::createMockDag1(node->getConfig().chain.dag_genesis_block.getHash().toString());
 
   for (int i = 1; i <= 9; i++) {
-    node->getDagManager()->addDagBlock(g_mock_dag0[i]);
+    node->getDagManager()->addDagBlock(std::move(g_mock_dag0[i]));
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply tests/full_node_test.cpp:760]: cause it is primitive type without any pointers so move does copy here anyway

File: tests/full_node_test.cpp [hunk 22]
```diff
@@ -906,7 +906,7 @@ TEST_F(FullNodeTest, reconstruct_dag) {
     // TODO: pbft does not support node stop yet, to be fixed ...
     node->getPbftManager()->stop();
     for (size_t i = 1; i < num_blks; i++) {
-      node->getDagManager()->addDagBlock(g_mock_dag0[i]);
+      node->getDagManager()->addDagBlock(std::move(mock_dags[i]));
```
Comment (MatusKysel) [inline tests/full_node_test.cpp:909]: now you are moving local variable 2x

Comment (JakubFornadel) [reply tests/full_node_test.cpp:909]: bad news is, second iteration of clang-tidy analysis did not find it...

File: libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp (hunk 13 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:29]: no, something is wrong with clang-format as even in our existing code base, some function are formatted with & like here, e.g. https://github.com/Taraxa-project/taraxa-node/blob/develop/libraries/core_libs/consensus/include/dag/dag.hpp#L66

and some are formatted with & right after the type(without whitespace), e.g. https://github.com/Taraxa-project/taraxa-node/blob/develop/libraries/core_libs/consensus/include/dag/sortition_params_manager.hpp#L24

Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/final_chain/replay_protection_service.hpp:29]: I created [issue](https://github.com/Taraxa-project/taraxa-node/issues/1662) for it so this PR can be merged
